// @u4/adbkit/dist/adb/thirdparty/STFService/STFServiceModel.js
"use strict";
// generarted by genProto.ts
Object.defineProperty(exports, "__esModule", { value: true });
exports.KeyEvent = exports.RingerMode = exports.ClipboardType = exports.MessageType = void 0;
var MessageType;
(function (MessageType) {
    MessageType[MessageType["DO_IDENTIFY"] = 1] = "DO_IDENTIFY";
    MessageType[MessageType["DO_KEYEVENT"] = 2] = "DO_KEYEVENT";
    MessageType[MessageType["DO_TYPE"] = 3] = "DO_TYPE";
    MessageType[MessageType["DO_WAKE"] = 4] = "DO_WAKE";
    MessageType[MessageType["DO_ADD_ACCOUNT_MENU"] = 24] = "DO_ADD_ACCOUNT_MENU";
    MessageType[MessageType["DO_REMOVE_ACCOUNT"] = 20] = "DO_REMOVE_ACCOUNT";
    MessageType[MessageType["GET_ACCOUNTS"] = 26] = "GET_ACCOUNTS";
    MessageType[MessageType["GET_BROWSERS"] = 5] = "GET_BROWSERS";
    MessageType[MessageType["GET_CLIPBOARD"] = 6] = "GET_CLIPBOARD";
    MessageType[MessageType["GET_DISPLAY"] = 19] = "GET_DISPLAY";
    MessageType[MessageType["GET_PROPERTIES"] = 7] = "GET_PROPERTIES";
    MessageType[MessageType["GET_RINGER_MODE"] = 27] = "GET_RINGER_MODE";
    MessageType[MessageType["GET_SD_STATUS"] = 25] = "GET_SD_STATUS";
    MessageType[MessageType["GET_VERSION"] = 8] = "GET_VERSION";
    MessageType[MessageType["GET_WIFI_STATUS"] = 23] = "GET_WIFI_STATUS";
    MessageType[MessageType["GET_BLUETOOTH_STATUS"] = 29] = "GET_BLUETOOTH_STATUS";
    MessageType[MessageType["GET_ROOT_STATUS"] = 31] = "GET_ROOT_STATUS";
    MessageType[MessageType["SET_CLIPBOARD"] = 9] = "SET_CLIPBOARD";
    MessageType[MessageType["SET_KEYGUARD_STATE"] = 10] = "SET_KEYGUARD_STATE";
    MessageType[MessageType["SET_RINGER_MODE"] = 21] = "SET_RINGER_MODE";
    MessageType[MessageType["SET_ROTATION"] = 12] = "SET_ROTATION";
    MessageType[MessageType["SET_WAKE_LOCK"] = 11] = "SET_WAKE_LOCK";
    MessageType[MessageType["SET_WIFI_ENABLED"] = 22] = "SET_WIFI_ENABLED";
    MessageType[MessageType["SET_BLUETOOTH_ENABLED"] = 30] = "SET_BLUETOOTH_ENABLED";
    MessageType[MessageType["SET_MASTER_MUTE"] = 28] = "SET_MASTER_MUTE";
    MessageType[MessageType["EVENT_AIRPLANE_MODE"] = 13] = "EVENT_AIRPLANE_MODE";
    MessageType[MessageType["EVENT_BATTERY"] = 14] = "EVENT_BATTERY";
    MessageType[MessageType["EVENT_CONNECTIVITY"] = 15] = "EVENT_CONNECTIVITY";
    MessageType[MessageType["EVENT_PHONE_STATE"] = 16] = "EVENT_PHONE_STATE";
    MessageType[MessageType["EVENT_ROTATION"] = 17] = "EVENT_ROTATION";
    MessageType[MessageType["EVENT_BROWSER_PACKAGE"] = 18] = "EVENT_BROWSER_PACKAGE";
})(MessageType = exports.MessageType || (exports.MessageType = {}));
var ClipboardType;
(function (ClipboardType) {
    ClipboardType[ClipboardType["TEXT"] = 1] = "TEXT";
})(ClipboardType = exports.ClipboardType || (exports.ClipboardType = {}));
var RingerMode;
(function (RingerMode) {
    RingerMode[RingerMode["SILENT"] = 0] = "SILENT";
    RingerMode[RingerMode["VIBRATE"] = 1] = "VIBRATE";
    RingerMode[RingerMode["NORMAL"] = 2] = "NORMAL";
})(RingerMode = exports.RingerMode || (exports.RingerMode = {}));
// Agent
var KeyEvent;
(function (KeyEvent) {
    KeyEvent[KeyEvent["DOWN"] = 0] = "DOWN";
    KeyEvent[KeyEvent["UP"] = 1] = "UP";
    KeyEvent[KeyEvent["PRESS"] = 2] = "PRESS";
})(KeyEvent = exports.KeyEvent || (exports.KeyEvent = {}));
