// @u4/adbkit/dist/adb/command/host-transport/Parcel.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ParcelVal = void 0;
/**
 * https://android.googlesource.com/platform/frameworks/native/+/master/libs/binder/ParcelValTypes.h
 */
var ParcelVal;
(function (ParcelVal) {
    ParcelVal[ParcelVal["VAL_NULL"] = -1] = "VAL_NULL";
    ParcelVal[ParcelVal["VAL_STRING"] = 0] = "VAL_STRING";
    ParcelVal[ParcelVal["VAL_INTEGER"] = 1] = "VAL_INTEGER";
    ParcelVal[ParcelVal["VAL_MAP"] = 2] = "VAL_MAP";
    ParcelVal[ParcelVal["VAL_BUNDLE"] = 3] = "VAL_BUNDLE";
    ParcelVal[ParcelVal["VAL_PARCELABLE"] = 4] = "VAL_PARCELABLE";
    ParcelVal[ParcelVal["VAL_SHORT"] = 5] = "VAL_SHORT";
    ParcelVal[ParcelVal["VAL_LONG"] = 6] = "VAL_LONG";
    ParcelVal[ParcelVal["VAL_FLOAT"] = 7] = "VAL_FLOAT";
    ParcelVal[ParcelVal["VAL_DOUBLE"] = 8] = "VAL_DOUBLE";
    ParcelVal[ParcelVal["VAL_BOOLEAN"] = 9] = "VAL_BOOLEAN";
    ParcelVal[ParcelVal["VAL_CHARSEQUENCE"] = 10] = "VAL_CHARSEQUENCE";
    ParcelVal[ParcelVal["VAL_LIST"] = 11] = "VAL_LIST";
    ParcelVal[ParcelVal["VAL_SPARSEARRAY"] = 12] = "VAL_SPARSEARRAY";
    ParcelVal[ParcelVal["VAL_BYTEARRAY"] = 13] = "VAL_BYTEARRAY";
    ParcelVal[ParcelVal["VAL_STRINGARRAY"] = 14] = "VAL_STRINGARRAY";
    ParcelVal[ParcelVal["VAL_IBINDER"] = 15] = "VAL_IBINDER";
    ParcelVal[ParcelVal["VAL_PARCELABLEARRAY"] = 16] = "VAL_PARCELABLEARRAY";
    ParcelVal[ParcelVal["VAL_OBJECTARRAY"] = 17] = "VAL_OBJECTARRAY";
    ParcelVal[ParcelVal["VAL_INTARRAY"] = 18] = "VAL_INTARRAY";
    ParcelVal[ParcelVal["VAL_LONGARRAY"] = 19] = "VAL_LONGARRAY";
    ParcelVal[ParcelVal["VAL_BYTE"] = 20] = "VAL_BYTE";
    ParcelVal[ParcelVal["VAL_SERIALIZABLE"] = 21] = "VAL_SERIALIZABLE";
    ParcelVal[ParcelVal["VAL_SPARSEBOOLEANARRAY"] = 22] = "VAL_SPARSEBOOLEANARRAY";
    ParcelVal[ParcelVal["VAL_BOOLEANARRAY"] = 23] = "VAL_BOOLEANARRAY";
    ParcelVal[ParcelVal["VAL_CHARSEQUENCEARRAY"] = 24] = "VAL_CHARSEQUENCEARRAY";
    ParcelVal[ParcelVal["VAL_PERSISTABLEBUNDLE"] = 25] = "VAL_PERSISTABLEBUNDLE";
    ParcelVal[ParcelVal["VAL_SIZE"] = 26] = "VAL_SIZE";
    ParcelVal[ParcelVal["VAL_SIZEF"] = 27] = "VAL_SIZEF";
    ParcelVal[ParcelVal["VAL_DOUBLEARRAY"] = 28] = "VAL_DOUBLEARRAY";
    ParcelVal[ParcelVal["VAL_CHAR"] = 29] = "VAL_CHAR";
    ParcelVal[ParcelVal["VAL_SHORTARRAY"] = 30] = "VAL_SHORTARRAY";
    ParcelVal[ParcelVal["VAL_CHARARRAY"] = 31] = "VAL_CHARARRAY";
    ParcelVal[ParcelVal["VAL_FLOATARRAY"] = 32] = "VAL_FLOATARRAY";
})(ParcelVal = exports.ParcelVal || (exports.ParcelVal = {}));
