// @u4/adbkit/dist/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.STFServiceModel = exports.Adb = exports.DevicePackage = exports.STFServiceBuf = exports.STFService = exports.Minicap = exports.Scrcpy = exports.Utils = exports.ParcelReader = exports.ServiceCallCommand = exports.IpRouteEntry = exports.IpRuleEntry = exports.ShellCommand = exports.KeyCodes = exports.KeyEvent = exports.ProcStat = exports.Stats = exports.Stats64 = exports.Entry = exports.Entry64 = exports.Pushtransfer = exports.PullTransfer = exports.Sync = exports.JdwpTracker = exports.DeviceClientExtra = exports.Tracker = exports.TcpUsbServer = exports.Connection = exports.DeviceClient = exports.Client = exports.AdbError = exports.AdbFailError = exports.AdbPrematureEOFError = exports.AdbUnexpectedDataError = exports.Parser = exports.MagicError = exports.ChecksumError = exports.PacketReader = exports.UnauthorizedError = exports.AuthError = exports.Socket = exports.Packet = exports.RollingCounter = exports.ServiceMap = exports.LateTransportError = exports.PrematurePacketError = exports.Service = exports.createClient = void 0;
var adb_1 = require("./adb");
Object.defineProperty(exports, "createClient", { enumerable: true, get: function () { return adb_1.createClient; } });
var service_1 = require("./adb/tcpusb/service");
Object.defineProperty(exports, "Service", { enumerable: true, get: function () { return __importDefault(service_1).default; } });
Object.defineProperty(exports, "PrematurePacketError", { enumerable: true, get: function () { return service_1.PrematurePacketError; } });
Object.defineProperty(exports, "LateTransportError", { enumerable: true, get: function () { return service_1.LateTransportError; } });
var servicemap_1 = require("./adb/tcpusb/servicemap");
Object.defineProperty(exports, "ServiceMap", { enumerable: true, get: function () { return __importDefault(servicemap_1).default; } });
var rollingcounter_1 = require("./adb/tcpusb/rollingcounter");
Object.defineProperty(exports, "RollingCounter", { enumerable: true, get: function () { return __importDefault(rollingcounter_1).default; } });
var packet_1 = require("./adb/tcpusb/packet");
Object.defineProperty(exports, "Packet", { enumerable: true, get: function () { return __importDefault(packet_1).default; } });
var socket_1 = require("./adb/tcpusb/socket");
Object.defineProperty(exports, "Socket", { enumerable: true, get: function () { return __importDefault(socket_1).default; } });
Object.defineProperty(exports, "AuthError", { enumerable: true, get: function () { return socket_1.AuthError; } });
Object.defineProperty(exports, "UnauthorizedError", { enumerable: true, get: function () { return socket_1.UnauthorizedError; } });
var packetreader_1 = require("./adb/tcpusb/packetreader");
Object.defineProperty(exports, "PacketReader", { enumerable: true, get: function () { return __importDefault(packetreader_1).default; } });
Object.defineProperty(exports, "ChecksumError", { enumerable: true, get: function () { return packetreader_1.ChecksumError; } });
Object.defineProperty(exports, "MagicError", { enumerable: true, get: function () { return packetreader_1.MagicError; } });
var parser_1 = require("./adb/parser");
Object.defineProperty(exports, "Parser", { enumerable: true, get: function () { return __importDefault(parser_1).default; } });
var errors_1 = require("./adb/errors");
Object.defineProperty(exports, "AdbUnexpectedDataError", { enumerable: true, get: function () { return errors_1.AdbUnexpectedDataError; } });
Object.defineProperty(exports, "AdbPrematureEOFError", { enumerable: true, get: function () { return errors_1.AdbPrematureEOFError; } });
Object.defineProperty(exports, "AdbFailError", { enumerable: true, get: function () { return errors_1.AdbFailError; } });
Object.defineProperty(exports, "AdbError", { enumerable: true, get: function () { return errors_1.AdbError; } });
var client_1 = require("./adb/client");
Object.defineProperty(exports, "Client", { enumerable: true, get: function () { return __importDefault(client_1).default; } });
var DeviceClient_1 = require("./adb/DeviceClient");
Object.defineProperty(exports, "DeviceClient", { enumerable: true, get: function () { return __importDefault(DeviceClient_1).default; } });
var connection_1 = require("./adb/connection");
Object.defineProperty(exports, "Connection", { enumerable: true, get: function () { return __importDefault(connection_1).default; } });
var server_1 = require("./adb/tcpusb/server");
Object.defineProperty(exports, "TcpUsbServer", { enumerable: true, get: function () { return __importDefault(server_1).default; } });
var tracker_1 = require("./adb/tracker");
Object.defineProperty(exports, "Tracker", { enumerable: true, get: function () { return __importDefault(tracker_1).default; } });
var DeviceClientExtra_1 = require("./adb/DeviceClientExtra");
Object.defineProperty(exports, "DeviceClientExtra", { enumerable: true, get: function () { return __importDefault(DeviceClientExtra_1).default; } });
var jdwptracker_1 = require("./adb/jdwptracker");
Object.defineProperty(exports, "JdwpTracker", { enumerable: true, get: function () { return __importDefault(jdwptracker_1).default; } });
var sync_1 = require("./adb/sync");
Object.defineProperty(exports, "Sync", { enumerable: true, get: function () { return __importDefault(sync_1).default; } });
var pulltransfer_1 = require("./adb/sync/pulltransfer");
Object.defineProperty(exports, "PullTransfer", { enumerable: true, get: function () { return __importDefault(pulltransfer_1).default; } });
var pushtransfer_1 = require("./adb/sync/pushtransfer");
Object.defineProperty(exports, "Pushtransfer", { enumerable: true, get: function () { return __importDefault(pushtransfer_1).default; } });
var entry64_1 = require("./adb/sync/entry64");
Object.defineProperty(exports, "Entry64", { enumerable: true, get: function () { return __importDefault(entry64_1).default; } });
var entry_1 = require("./adb/sync/entry");
Object.defineProperty(exports, "Entry", { enumerable: true, get: function () { return __importDefault(entry_1).default; } });
var stats64_1 = require("./adb/sync/stats64");
Object.defineProperty(exports, "Stats64", { enumerable: true, get: function () { return __importDefault(stats64_1).default; } });
var stats_1 = require("./adb/sync/stats");
Object.defineProperty(exports, "Stats", { enumerable: true, get: function () { return __importDefault(stats_1).default; } });
var stat_1 = require("./adb/proc/stat");
Object.defineProperty(exports, "ProcStat", { enumerable: true, get: function () { return __importDefault(stat_1).default; } });
var STFServiceModel_1 = require("./adb/thirdparty/STFService/STFServiceModel");
Object.defineProperty(exports, "KeyEvent", { enumerable: true, get: function () { return STFServiceModel_1.KeyEvent; } });
// export android key enumeration
var keycode_1 = require("./adb/keycode");
Object.defineProperty(exports, "KeyCodes", { enumerable: true, get: function () { return keycode_1.KeyCodes; } });
var host_transport_1 = require("./adb/command/host-transport/");
Object.defineProperty(exports, "ShellCommand", { enumerable: true, get: function () { return host_transport_1.ShellCommand; } });
Object.defineProperty(exports, "IpRuleEntry", { enumerable: true, get: function () { return host_transport_1.IpRuleEntry; } });
Object.defineProperty(exports, "IpRouteEntry", { enumerable: true, get: function () { return host_transport_1.IpRouteEntry; } });
var serviceCall_1 = require("./adb/command/host-transport/serviceCall");
Object.defineProperty(exports, "ServiceCallCommand", { enumerable: true, get: function () { return __importDefault(serviceCall_1).default; } });
Object.defineProperty(exports, "ParcelReader", { enumerable: true, get: function () { return serviceCall_1.ParcelReader; } });
// give access to Utils class ( readAll and parsePublicKey)
var utils_1 = require("./adb/utils");
Object.defineProperty(exports, "Utils", { enumerable: true, get: function () { return __importDefault(utils_1).default; } });
var Scrcpy_1 = require("./adb/thirdparty/scrcpy/Scrcpy");
Object.defineProperty(exports, "Scrcpy", { enumerable: true, get: function () { return __importDefault(Scrcpy_1).default; } });
var Minicap_1 = require("./adb/thirdparty/minicap/Minicap");
Object.defineProperty(exports, "Minicap", { enumerable: true, get: function () { return __importDefault(Minicap_1).default; } });
var STFService_1 = require("./adb/thirdparty/STFService/STFService");
Object.defineProperty(exports, "STFService", { enumerable: true, get: function () { return __importDefault(STFService_1).default; } });
var STFServiceBuf_1 = require("./adb/thirdparty/STFService/STFServiceBuf");
Object.defineProperty(exports, "STFServiceBuf", { enumerable: true, get: function () { return __importDefault(STFServiceBuf_1).default; } });
var DevicePackage_1 = require("./adb/DevicePackage");
Object.defineProperty(exports, "DevicePackage", { enumerable: true, get: function () { return __importDefault(DevicePackage_1).default; } });
/**
 * main entry point
 */
const adb_2 = require("./adb");
const utils_2 = __importDefault(require("./adb/utils"));
/**
 * Keep @u4/adbkit v3.x old adb export
 */
exports.Adb = {
    util: utils_2.default,
    createClient: adb_2.createClient
};
exports.STFServiceModel = __importStar(require("./adb/thirdparty/STFService/STFServiceModel"));
/**
 * Keep @u4/adbkit v3.x default export shape
 */
exports.default = exports.Adb;
