// @u4/adbkit/dist/adb/sync/stats64.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const stats_1 = __importDefault(require("./stats"));
const b1k = BigInt(1000);
class Stats64 {
    isFile() {
        return !!(Number(this.mode) & stats_1.default.S_IFREG);
    }
    isDirectory() {
        return !!(Number(this.mode) & stats_1.default.S_IFDIR);
    }
    isBlockDevice() {
        return !!(Number(this.mode) & stats_1.default.S_IFBLK);
    }
    isCharacterDevice() {
        return !!(Number(this.mode) & stats_1.default.S_IFCHR);
    }
    isSymbolicLink() {
        return !!(Number(this.mode) & stats_1.default.S_IFLNK);
    }
    isFIFO() {
        return !!(Number(this.mode) & stats_1.default.S_IFIFO);
    }
    isSocket() {
        return !!(Number(this.mode) & stats_1.default.S_IFSOCK);
    }
    get atimeMs() {
        return this.atimeNs / b1k;
    }
    get mtimeMs() {
        return this.mtimeNs / b1k;
    }
    get ctimeMs() {
        return this.ctimeNs / b1k;
    }
    get birthtimeMs() {
        return this.birthtimeMs / b1k;
    }
    get atime() {
        return new Date(Number(this.atimeMs));
    }
    get mtime() {
        return new Date(Number(this.mtimeMs));
    }
    get ctime() {
        return new Date(Number(this.ctimeMs));
    }
    get birthtimeNs() {
        return this.ctimeNs;
    }
    get birthtime() {
        return this.ctime;
    }
    constructor(error, dev, ino, mode, nlink, uid, gid, size, atimeNs, mtimeNs, ctimeNs) {
        this.error = error;
        this.dev = dev;
        this.ino = ino;
        this.mode = mode;
        this.nlink = nlink;
        this.uid = uid;
        this.gid = gid;
        this.size = size;
        this.atimeNs = atimeNs;
        this.mtimeNs = mtimeNs;
        this.ctimeNs = ctimeNs;
        this.rdev = BigInt(0);
        this.blksize = BigInt(0);
        this.blocks = BigInt(0);
    }
}
exports.default = Stats64;
